components/weekly_storage/weekly_event_storage.h

```diff
@@ -0,0 +1,136 @@
+/* Copyright 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_WEEKLY_STORAGE_WEEKLY_EVENT_STORAGE_H_
+#define BRAVE_COMPONENTS_WEEKLY_STORAGE_WEEKLY_EVENT_STORAGE_H_
+
+#include <list>
+#include <memory>
+#include <utility>
+
+#include "base/time/clock.h"
+#include "base/time/default_clock.h"
+#include "base/time/time.h"
+#include "base/values.h"
+#include "components/prefs/pref_service.h"
+#include "components/prefs/scoped_user_pref_update.h"
+#include "third_party/abseil-cpp/absl/types/optional.h"
+
+// WeeklyStorage variant holding a list of events over the past week.
+//
+// Mostly used by various P3A recorders to report whether an event happened
+// during the measurement period.
+//
+// New event values are recorded by calling `Add()` and are forgotten
+// after approximately a week.
+//
+// Parametarized over an `enum class` to get some type checking for the caller.
+// Requires |pref_name| to be already registered.
+template <typename T>
+class WeeklyEventStorage {
+ public:
+  WeeklyEventStorage(PrefService* prefs, const char* pref_name)
+      : prefs_(prefs),
+        pref_name_(pref_name),
+        clock_(std::make_unique<base::DefaultClock>()) {
+    DCHECK(pref_name);
+    if (prefs) {
+      Load();
+    }
+  }
+
+  // Accept an explicit clock so tests can manipulate the passage of time.
+  WeeklyEventStorage(PrefService* prefs,
+                     const char* pref_name,
+                     std::unique_ptr<base::Clock> clock)
+      : prefs_(prefs), pref_name_(pref_name), clock_(std::move(clock)) {
+    DCHECK(prefs);
+    DCHECK(pref_name);
+    Load();
+  }
+
+  ~WeeklyEventStorage() = default;
+
+  WeeklyEventStorage(const WeeklyEventStorage&) = delete;
+  WeeklyEventStorage& operator=(const WeeklyEventStorage&) = delete;
+
+  void Add(T value) {
+    FilterToWeek();
+    // Round the timestamp to the nearest day to make correlation harder.
+    base::Time day = clock_->Now().LocalMidnight();
+    events_.push_front({day, value});
+    Save();
+  }
+
+  absl::optional<T> GetLatest() {
+    auto result = absl::optional<T>();
+    if (HasEvent()) {
+      // Assume the front is the most recent event.
+      result = events_.front().value;
```

Comment:
maybe I'm missing something, but why do we keep all events, but are using only the first one?